core/types/hashing.go

```diff
@@ -89,6 +89,7 @@ func DeriveSha(list DerivableList, hasher TrieHasher) common.Hash {
 
 	valueBuf := encodeBufferPool.Get().(*bytes.Buffer)
 	defer encodeBufferPool.Put(valueBuf)
+	defer valueBuf.Reset()
```

Comment:
`byte.Buffer` will still retain the underlying storage even Reset is called.

Reply:
Yes, `byte.Buffer` will return to `sync.Pool` and retains the underlying storage for use by future writes, but `Reset()` will do `b.buf = b.buf[:0]`, so it can release the unnecessary memory:

https://github.com/golang/go/blob/master/src/bytes/buffer.go#L94-L101